### ISSUE TITLE
Allow simple rule instead of rulebase file for mmnormalize

### ISSUE THREAD

**Issue by hakman** [rsyslog/rsyslog#625]
RuleBase files make config files less readable.
When providing examples to customers or when reviewing configs, it would make things easier to have a single config file that can also inculde the rule as a string (instead of a file name and a second config file).

In most cases, string containing a simple rule would be enough.
If possible, to have more rules in some sort of array, that would work nicely also.

**Comment by mostolog** [comment]
Do you have a scheduled milestone for this? I have a bunch of directories to hold .conf+.rb instead of a bunch of files...

**Comment by rgerhards** [comment]
I'll tenatively schedule it for 8.24, but at low prio. Need to see how much we actually need to change.

**Comment by davidelang** [comment]
While I am not opposed to this, I question how valuable it really is.

tl;dr I think that if you have more than a few calls to mmnormalize, you are probably doing something wrong. And if you are doing much of anything with it, your ruleset is going to be long enough to be very ugly to put in rsyslog.conf

In my config, I would possibly use it as I use mmnormalize instead of mmjsonarse because I have broken senders that send json that mmjsonarse can't handle, so I need a very simple (4-line) rule to handle this. So I have three rulesets, one to try and fix the rawmsg, one if the message format is json, and then one to deal with everything once I've gotten the framing fixed from the first two.

But I don't see why people would have a lot of different .rb files. I suspect it's admin inertia from experience with regular expression parsers, where having unrelated things in one file is a problem, so you have a lot of high-level logic to look at the message contents and then figure out which rules to apply to it.

With liblognorm this is very seldom needed. If you can combine the programname with the message and include the programname in your rules (prefix works wonders here), liblognorm is extremely efficient in only using the relevant rules.

We even have an include option so that you can have your rules as separate files for ease of administering them, but then have a single ruleset as far as rsyslog is concerned.

**Comment by rgerhards** [comment]
@davidelang I agree to most of your comments, using different rulesets usually leads to lower performance. However, there may be real simple cases where it could be useful. And, more important, this is staging work for things like pulling imfile metadata...

So while I think the use case might not be strong, there is value in supporting the functionality. Especially as I think/hope it's not a big task to do.

**Comment by davidelang** [comment]
> So while I think the use case might not be strong, there is value in supporting the functionality. Especially as I think/hope it's not a big task to do.

I agree, which is why I started off my comment by saying I'm not opposed to it.

But I want the people who are looking for it to think a bit about what they are 
doing and see if they should change things.

David Lang

**Comment by mostolog** [comment]
Thanks a lot @davidelang @rgerhards for your comments, which occasionally led me to discover rsyslog _hidden gems_ :stuck_out_tongue: 

>We even have an include option so that you can have your rules as separate files for ease of administering them, but then have a single ruleset as far as rsyslog is concerned.

In our scenario we have MANY grok/regex patterns, which made me think how to specify multiple files. While testing, I was "bored" to have separated files and doing copy-paste. On the other hand, above quoted paragraph is fair enough to do what we need. I'll keep different .rb files and include them in our config.
Is the include documented somewhere? Isn't this a rsyslog feature but a liblognorm's ?

**Comment by mostolog** [comment]
Considering our scenario, multiple apps, each one with his own format+pipeline, this still makes sense in order to have a single app.conf file (and include it), instead of having 2 files (conf+rb)
Isn't it?

**Comment by davidelang** [comment]
http://www.liblognorm.com/files/manual/configuration.html

also take a look at Rainer's paper comparing Grok to liblognorm, https://www.fernuni-hagen.de/imperia/md/content/rechnerarchitektur/rainer_gerhards.pdf

As far as your scenario, I would still have all the logs going to one place, and get split from there. That one central place can parse the logs before they are distributed to the per-app pipelines. It is a single point of load, but it's less CPU used overall.

**Comment by rgerhards** [comment]
2016-11-23 12:26 GMT+01:00 mostolog <notifications@github.com>:

> Considering our scenario, multiple apps, each one with his own
> format+pipeline, this still makes sense in order to have a single app.conf
> file (and include it), instead of having 2 files (conf+rb)
>

Short answer: Yup, single file makes sense.

Long answer: read this:

https://www.researchgate.net/publication/310545144_Efficient_Normalization_of_IT_Log_Messages_under_Realtime_Conditions

Rainer

**Comment by mostolog** [comment]
A few days ago we talked about how liblognorm works. Now I know it uses (predictive) trees.
I'm awaiting this feature in order to move my _logstash.conf_ files to _rsyslog.conf_ (rules written within the same file).

Thanks!

**Comment by rgerhards** [comment]
small nit: lognorm is not using trees any longer, it uses a (disconnected) DAG. But in this discussion, it really doesn't matter, the understanding of a tree is well enough.

**Comment by rgerhards** [comment]
implemented via https://github.com/rsyslog/rsyslog/pull/1380
